File: internal/notify/notify.go
package notify

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"os/exec"
	"runtime"
	"strings"
	"time"
)

// Event describes a run notification payload.
type Event struct {
	Name           string
	Source         string
	Status         string
	ExitCode       int
	Duration       time.Duration
	StartedAt      time.Time
	CommandPreview string
	StderrTail     string
}

// Manager dispatches notifications to configured providers.
type Manager struct {
	DesktopEnabled bool
	WebhookURL     string
	WebhookTimeout time.Duration
	HTTPClient     *http.Client
}

// Notify sends an event to all configured providers.
func (m Manager) Notify(ctx context.Context, e Event) error {
	var errs []string

	if m.DesktopEnabled {
		if err := desktopNotify(ctx, e); err != nil {
			errs = append(errs, fmt.Sprintf("desktop: %v", err))
		}
	}

	if m.WebhookURL != "" {
		if err := webhookNotify(ctx, m.WebhookURL, m.WebhookTimeout, m.HTTPClient, e); err != nil {
			errs = append(errs, fmt.Sprintf("webhook: %v", err))
		}
	}

	if len(errs) > 0 {
		return fmt.Errorf("%s", strings.Join(errs, "; "))
	}

	return nil
}

func desktopNotify(ctx context.Context, e Event) error {
	title := fmt.Sprintf("otto: %s %s", e.Name, e.Status)
	body := fmt.Sprintf("exit %d, duration %s", e.ExitCode, e.Duration.Round(time.Millisecond))

	switch runtime.GOOS {
	case "darwin":
		script := fmt.Sprintf(`display notification %q with title %q`, body, title)
		return exec.CommandContext(ctx, "osascript", "-e", script).Run()
	case "linux":
		return exec.CommandContext(ctx, "notify-send", title, body).Run()
	default:
		return nil
	}
}

func webhookNotify(ctx context.Context, webhookURL string, timeout time.Duration, client *http.Client, e Event) error {
	if timeout <= 0 {
		timeout = 5 * time.Second
	}

	if client == nil {
		client = &http.Client{Timeout: timeout}
	}

	payload := map[string]any{
		"name":            e.Name,
		"source":          e.Source,
		"status":          e.Status,
		"exit_code":       e.ExitCode,
		"duration_ms":     e.Duration.Milliseconds(),
		"started_at":      e.StartedAt.UTC().Format(time.RFC3339),
		"command_preview": e.CommandPreview,
		"stderr_tail":     e.StderrTail,
	}

	b, err := json.Marshal(payload)
	if err != nil {
		return fmt.Errorf("marshal payload: %w", err)
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, webhookURL, bytes.NewReader(b))
	if err != nil {
		return fmt.Errorf("build request: %w", err)
	}

	req.Header.Set("Content-Type", "application/json")

	resp, err := client.Do(req)
	if err != nil {
		return fmt.Errorf("send request: %w", err)
	}

	defer func() {
		_ = resp.Body.Close()
	}()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return fmt.Errorf("unexpected status %d", resp.StatusCode)
	}

	return nil
}

File: src/notify.rs
use reqwest::blocking::Client;
use serde::Serialize;
use std::process::Command;
use std::time::Duration;
use time::OffsetDateTime;

#[derive(Debug, Clone)]
pub struct Event {
    pub name: String,
    pub source: String,
    pub status: String,
    pub exit_code: i32,
    pub duration: Duration,
    pub started_at: OffsetDateTime,
    pub command_preview: String,
    pub stderr_tail: Option<String>,
}

#[derive(Debug, Clone)]
pub struct Manager {
    pub desktop_enabled: bool,
    pub webhook_url: String,
    pub webhook_timeout: Duration,
}

impl Manager {
    pub fn notify(&self, event: &Event) -> Result<(), String> {
        let mut errors = Vec::new();

        if self.desktop_enabled
            && let Err(err) = desktop_notify(event)
        {
            errors.push(format!("desktop: {err}"));
        }

        if !self.webhook_url.is_empty()
            && let Err(err) = webhook_notify(&self.webhook_url, self.webhook_timeout, event)
        {
            errors.push(format!("webhook: {err}"));
        }

        if errors.is_empty() {
            Ok(())
        } else {
            Err(errors.join("; "))
        }
    }
}

fn desktop_notify(event: &Event) -> Result<(), String> {
    let title = format!("{} {}", event.name, event.status);
    let body = format!(
        "exit {}, duration {}",
        event.exit_code,
        format_duration(event.duration)
    );

    if cfg!(target_os = "macos") {
        let script = format!("display notification {:?} with title {:?}", body, title);

        let status = Command::new("osascript")
            .arg("-e")
            .arg(script)
            .status()
            .map_err(|e| e.to_string())?;

        if status.success() {
            return Ok(());
        }

        return Err(format!("osascript exited with status {status}"));
    }

    if cfg!(target_os = "linux") {
        let status = Command::new("notify-send")
            .arg(&title)
            .arg(&body)
            .status()
            .map_err(|e| e.to_string())?;

        if status.success() {
            return Ok(());
        }

        return Err(format!("notify-send exited with status {status}"));
    }

    Ok(())
}

#[derive(Debug, Serialize)]
struct WebhookPayload<'a> {
    name: &'a str,
    source: &'a str,
    status: &'a str,
    exit_code: i32,
    duration_ms: i128,
    started_at: String,
    command_preview: &'a str,
    stderr_tail: &'a str,
}

fn webhook_notify(webhook_url: &str, timeout: Duration, event: &Event) -> Result<(), String> {
    let timeout = if timeout.is_zero() {
        Duration::from_secs(5)
    } else {
        timeout
    };

    let client = Client::builder()
        .timeout(timeout)
        .build()
        .map_err(|e| format!("build client: {e}"))?;

    let payload = WebhookPayload {
        name: &event.name,
        source: &event.source,
        status: &event.status,
        exit_code: event.exit_code,
        duration_ms: event.duration.as_millis() as i128,
        started_at: event
            .started_at
            .format(&time::format_description::well_known::Rfc3339)
            .map_err(|e| format!("format started_at: {e}"))?,
        command_preview: &event.command_preview,
        stderr_tail: event.stderr_tail.as_deref().unwrap_or(""),
    };

    let response = client
        .post(webhook_url)
        .header(reqwest::header::CONTENT_TYPE, "application/json")
        .json(&payload)
        .send()
        .map_err(|e| format!("send request: {e}"))?;

    if response.status().is_success() {
        Ok(())
    } else {
        Err(format!("unexpected status {}", response.status().as_u16()))
    }
}

fn format_duration(duration: Duration) -> String {
    let ms = duration.as_millis();
    if ms < 1_000 {
        return format!("{ms}ms");
    }

    if ms.is_multiple_of(1_000) {
        return format!("{}s", ms / 1_000);
    }

    format!("{:.3}s", duration.as_secs_f64())
}
